File: src/util/paths.ts
import {
  appendFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, join } from "node:path";

function home(): string {
  return homedir();
}

// ─── Claude Code ─────────────────────────────────────────────

export interface ClaudePaths {
  dir: string;
  globalJson: string;
  settings: string;
  skillsDir: string;
  agentsMd: string;
}

export function claudePaths(): ClaudePaths {
  const h = home();
  const dir = join(h, ".claude");
  return {
    dir,
    globalJson: join(h, ".claude.json"),
    settings: join(dir, "settings.json"),
    skillsDir: join(dir, "skills"),
    agentsMd: join(dir, "AGENTS.md"),
  };
}

export function claudeKnownBinDirs(): string[] {
  return [join(home(), ".local", "bin")];
}

export function claudeDesktopPaths(): string[] {
  if (process.platform === "win32") {
    const paths: string[] = [];
    const local = process.env.LOCALAPPDATA;
    if (local) paths.push(join(local, "AnthropicClaude", "claude.exe"));
    const roam = process.env.APPDATA;
    if (roam) paths.push(join(roam, "Claude", "claude.exe"));
    return paths;
  }
  if (process.platform === "darwin") {
    return ["/Applications/Claude.app"];
  }
  return [];
}

// ─── OpenCode ────────────────────────────────────────────────

export interface OpenCodePaths {
  dir: string;
  config: string;
  agentsMd: string;
}

export function opencodePaths(): OpenCodePaths {
  let dir: string;
  if (process.platform === "win32") {
    const roam = process.env.APPDATA;
    dir = roam ? join(roam, "opencode") : join(home(), ".config", "opencode");
  } else {
    dir = join(home(), ".config", "opencode");
  }
  return {
    dir,
    config: join(dir, "config.json"),
    agentsMd: join(dir, "AGENTS.md"),
  };
}

export function opencodeKnownBinDirs(): string[] {
  const dirs = [join(home(), ".opencode", "bin"), join(home(), ".local", "bin")];
  if (process.platform === "win32") {
    dirs.push(join(home(), "scoop", "shims"));
  }
  return dirs;
}

export function opencodeDesktopPaths(): string[] {
  if (process.platform === "win32") {
    const local = process.env.LOCALAPPDATA;
    if (local) return [join(local, "Programs", "OpenCode", "OpenCode.exe")];
    return [];
  }
  if (process.platform === "darwin") {
    return ["/Applications/OpenCode.app"];
  }
  return ["/usr/bin/ai.opencode.desktop"];
}

// ─── Codex (OpenAI) ──────────────────────────────────────────

export interface CodexPaths {
  dir: string;
  config: string;
  hooks: string;
  instructions: string;
}

export function codexPaths(): CodexPaths {
  const dir = join(home(), ".codex");
  return {
    dir,
    config: join(dir, "config.toml"),
    hooks: join(dir, "hooks.json"),
    instructions: join(dir, "instructions.md"),
  };
}

export function codexKnownBinDirs(): string[] {
  const dirs: string[] = [];
  const envDir = process.env.CODEX_INSTALL_DIR;
  if (envDir) dirs.push(envDir);
  if (process.platform === "win32") {
    const la = process.env.LOCALAPPDATA;
    if (la) dirs.push(join(la, "Programs", "OpenAI", "Codex", "bin"));
  }
  dirs.push(join(home(), ".local", "bin"));
  dirs.push(join(home(), ".cargo", "bin"));
  return dirs;
}

// ─── Antigravity (Google) ────────────────────────────────────

export interface AntigravityPaths {
  dir: string;
  mcpCli: string;
  mcpIde: string;
  settingsCli: string;
  settingsIde: string;
  hooks: string;
  agentsMd: string;
}

export function antigravityPaths(): AntigravityPaths {
  const gemini = join(home(), ".gemini");
  return {
    dir: gemini,
    mcpCli: join(gemini, "antigravity-cli", "mcp_config.json"),
    mcpIde: join(gemini, "antigravity-ide", "mcp_config.json"),
    settingsCli: join(gemini, "antigravity-cli", "settings.json"),
    settingsIde: join(gemini, "antigravity-ide", "settings.json"),
    hooks: join(gemini, "config", "hooks.json"),
    agentsMd: join(gemini, "config", "AGENTS.md"),
  };
}

/** All MCP config files that Antigravity reads. */
export function antigravityMcpFiles(): string[] {
  const gemini = join(home(), ".gemini");
  const files = [
    join(gemini, "antigravity-cli", "mcp_config.json"),
    join(gemini, "antigravity-ide", "mcp_config.json"),
  ];
  const desktop = join(gemini, "antigravity-desktop");
  if (existsSync(desktop)) {
    files.push(join(desktop, "mcp_config.json"));
  }
  return files;
}

/** All settings.json files that need permissions entries. */
export function antigravitySettingsFiles(): string[] {
  const gemini = join(home(), ".gemini");
  return [
    join(gemini, "antigravity-cli", "settings.json"),
    join(gemini, "antigravity-ide", "settings.json"),
  ];
}

export function antigravityKnownBinDirs(): string[] {
  if (process.platform === "win32") {
    const local = process.env.LOCALAPPDATA;
    return local ? [join(local, "agy", "bin")] : [];
  }
  return [join(home(), ".local", "bin")];
}

export function antigravityDesktopPaths(): string[] {
  if (process.platform === "win32") {
    const local = process.env.LOCALAPPDATA;
    if (!local) return [];
    return [
      join(local, "Programs", "Antigravity", "Antigravity.exe"),
      join(local, "Programs", "Antigravity IDE", "Antigravity IDE.exe"),
    ];
  }
  if (process.platform === "darwin") {
    return ["/Applications/Antigravity.app", "/Applications/Antigravity IDE.app"];
  }
  return ["/opt/antigravity", "/opt/antigravity-ide"];
}

// ─── Shared ──────────────────────────────────────────────────

/** Local bin directory for tool installs. */
export function localBin(): string {
  if (process.platform === "win32") {
    const la = process.env.LOCALAPPDATA;
    if (la) return join(la, "Programs", "toksave");
    // Fallback when LOCALAPPDATA is not set
    return join(home(), "AppData", "Local", "Programs", "toksave");
  }
  return join(home(), ".local", "bin");
}

/** Cache directory for toksave. */
export function cacheDir(): string {
  if (process.env.TOKSAVE_CACHE_DIR) return process.env.TOKSAVE_CACHE_DIR;
  return join(home(), ".cache", "toksave");
}

/** Ensure a directory exists. */
export function ensureDir(path: string): void {
  if (!existsSync(path)) {
    mkdirSync(path, { recursive: true });
  }
}

/** Read a file, returning null if it doesn't exist. */
export function readFile(path: string): string | null {
  try {
    return readFileSync(path, "utf-8");
  } catch {
    return null;
  }
}

/** Write a file, creating parent dirs if needed. */
export function writeFile(path: string, content: string): void {
  ensureDir(dirname(path));
  const tmp = `${path}.${process.pid}.tmp`;
  writeFileSync(tmp, content, "utf-8");
  renameSync(tmp, path);
}

/** Append to a file, creating it if needed. */
export function appendFile(path: string, content: string): void {
  ensureDir(dirname(path));
  appendFileSync(path, content, "utf-8");
}

/** Get the toksave binary absolute path or alias. */
export function toksaveAbs(): string {
  if (process.env.NODE_ENV === "test") return "toksave";

  const candidates = [process.argv[1], process.execPath, join(localBin(), "toksave")];
  if (process.platform === "win32") candidates.push(join(localBin(), "toksave.exe"));

  for (const candidate of candidates) {
    if (!candidate || !existsSync(candidate)) continue;
    const name = basename(candidate).toLowerCase();
    if (name === "toksave" || name === "toksave.exe") return candidate;
  }

  return "toksave";
}

File: src/util/paths.rs
use crate::util::errors::Result;
use std::env;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

pub fn home() -> PathBuf {
    if let Some(h) = env::var_os("HOME") {
        return PathBuf::from(h);
    }
    if let Some(h) = env::var_os("USERPROFILE") {
        return PathBuf::from(h);
    }
    dirs::home_dir().unwrap_or_else(|| PathBuf::from("."))
}

pub struct ClaudePaths {
    pub dir: PathBuf,
    pub global_json: PathBuf,
    pub settings: PathBuf,
    pub skills_dir: PathBuf,
    pub agents_md: PathBuf,
}

pub fn claude_paths() -> ClaudePaths {
    let h = home();
    let dir = h.join(".claude");
    ClaudePaths {
        global_json: h.join(".claude.json"),
        settings: dir.join("settings.json"),
        skills_dir: dir.join("skills"),
        agents_md: dir.join("AGENTS.md"),
        dir,
    }
}

pub fn claude_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub fn claude_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        let mut v = vec![];
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            v.push(
                PathBuf::from(local)
                    .join("AnthropicClaude")
                    .join("claude.exe"),
            );
        }
        if let Some(roam) = env::var_os("APPDATA") {
            v.push(PathBuf::from(roam).join("Claude").join("claude.exe"));
        }
        v
    } else if cfg!(target_os = "macos") {
        vec![PathBuf::from("/Applications/Claude.app")]
    } else {
        vec![]
    }
}

pub struct OpencodePaths {
    pub dir: PathBuf,
    pub config: PathBuf,
    pub agents_md: PathBuf,
    pub plugins_dir: PathBuf,
}

pub fn opencode_paths() -> OpencodePaths {
    let h = home();
    let dir = h.join(".config").join("opencode");
    OpencodePaths {
        config: dir.join("config.json"),
        agents_md: dir.join("AGENTS.md"),
        plugins_dir: dir.join("plugins"),
        dir,
    }
}

pub fn opencode_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub fn opencode_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            return vec![
                PathBuf::from(local)
                    .join("Programs")
                    .join("OpenCode")
                    .join("OpenCode.exe"),
            ];
        }
        vec![]
    } else if cfg!(target_os = "macos") {
        vec![PathBuf::from("/Applications/OpenCode.app")]
    } else {
        vec![PathBuf::from("/usr/bin/ai.opencode.desktop")]
    }
}

pub struct CodexPaths {
    pub dir: PathBuf,
    pub config: PathBuf,
    pub hooks: PathBuf,
    pub instructions: PathBuf,
}

pub fn codex_paths() -> CodexPaths {
    let h = home();
    let dir = h.join(".codex");
    CodexPaths {
        config: dir.join("config.toml"),
        hooks: dir.join("hooks.json"),
        instructions: dir.join("instructions.md"),
        dir,
    }
}

pub fn codex_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub struct AntigravityPaths {
    pub dir: PathBuf,
    pub hooks: PathBuf,
}

pub fn antigravity_paths() -> AntigravityPaths {
    let h = home();
    let gemini = h.join(".gemini");
    let dir = if gemini.exists() {
        gemini
    } else {
        h.join(".antigravity")
    };
    AntigravityPaths {
        hooks: dir.join("config").join("hooks.json"),
        dir,
    }
}

pub fn antigravity_known_bin_dirs() -> Vec<PathBuf> {
    let mut v = vec![home().join(".local").join("bin")];
    if cfg!(windows)
        && let Some(local) = env::var_os("LOCALAPPDATA")
    {
        v.push(PathBuf::from(local).join("agy").join("bin"));
    }
    v
}

pub fn antigravity_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        let mut v = vec![];
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            let p = PathBuf::from(local);
            v.push(
                p.join("Programs")
                    .join("Antigravity")
                    .join("Antigravity.exe"),
            );
            v.push(
                p.join("Programs")
                    .join("Antigravity IDE")
                    .join("Antigravity IDE.exe"),
            );
        }
        v
    } else if cfg!(target_os = "macos") {
        vec![
            PathBuf::from("/Applications/Antigravity.app"),
            PathBuf::from("/Applications/Antigravity IDE.app"),
        ]
    } else {
        vec![
            PathBuf::from("/opt/antigravity"),
            PathBuf::from("/opt/antigravity-ide"),
        ]
    }
}

pub fn antigravity_mcp_files() -> Vec<PathBuf> {
    let p = antigravity_paths();
    vec![
        p.dir.join("mcp.json"),
        p.dir.join("config").join("mcp.json"),
    ]
}

pub fn antigravity_settings_files() -> Vec<PathBuf> {
    let p = antigravity_paths();
    vec![
        p.dir.join("settings.json"),
        p.dir.join("config").join("settings.json"),
    ]
}

pub struct CopilotPaths {
    pub dir: PathBuf,
    pub hooks_dir: PathBuf,
    pub mcp_config: PathBuf,
    pub instructions: PathBuf,
    pub skills_dir: PathBuf,
}

pub fn copilot_paths() -> CopilotPaths {
    let h = home();
    let dir = h.join(".copilot");
    CopilotPaths {
        hooks_dir: dir.join("hooks"),
        mcp_config: dir.join("mcp.json"),
        instructions: dir.join("instructions.md"),
        skills_dir: dir.join("skills"),
        dir,
    }
}

pub fn copilot_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub struct DroidPaths {
    pub dir: PathBuf,
    pub hooks_file: PathBuf,
    pub mcp_config: PathBuf,
}

pub fn droid_paths() -> DroidPaths {
    let h = home();
    let dir = h.join(".factory-droid");
    DroidPaths {
        hooks_file: dir.join("hooks.json"),
        mcp_config: dir.join("mcp.json"),
        dir,
    }
}

pub fn droid_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub fn droid_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            return vec![
                PathBuf::from(local)
                    .join("Programs")
                    .join("Factory Droid")
                    .join("Factory Droid.exe"),
            ];
        }
        vec![]
    } else if cfg!(target_os = "macos") {
        vec![PathBuf::from("/Applications/Factory Droid.app")]
    } else {
        vec![]
    }
}

pub struct DevinPaths {
    pub dir: PathBuf,
    pub hooks_file: PathBuf,
    pub mcp_config: PathBuf,
}

pub fn devin_paths() -> DevinPaths {
    let h = home();
    let dir = h.join(".devin");
    DevinPaths {
        hooks_file: dir.join("hooks.json"),
        mcp_config: dir.join("mcp.json"),
        dir,
    }
}

pub fn devin_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub fn devin_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            return vec![
                PathBuf::from(local)
                    .join("Programs")
                    .join("Devin")
                    .join("Devin.exe"),
            ];
        }
        vec![]
    } else if cfg!(target_os = "macos") {
        vec![PathBuf::from("/Applications/Devin.app")]
    } else {
        vec![]
    }
}

pub struct WarpPaths {
    pub dir: PathBuf,
    pub hooks_file: PathBuf,
    pub mcp_config: PathBuf,
    pub mcp_config_official: PathBuf,
}

pub fn warp_paths() -> WarpPaths {
    let h = home();
    let dir = h.join(".warp");
    WarpPaths {
        hooks_file: dir.join("hooks.json"),
        mcp_config: dir.join("mcp.json"),
        mcp_config_official: dir.join(".mcp.json"),
        dir,
    }
}

pub struct WarpCliPaths {
    pub dir: PathBuf,
    pub settings: PathBuf,
    pub mcp_config: PathBuf,
}

fn xdg_config_home() -> PathBuf {
    env::var_os("XDG_CONFIG_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| home().join(".config"))
}

/// Warp Agent CLI settings/MCP directory. Separate from the desktop app.
pub fn warp_cli_paths() -> WarpCliPaths {
    if cfg!(windows) {
        let dir = env::var_os("LOCALAPPDATA")
            .map(|l| {
                PathBuf::from(l)
                    .join("warp")
                    .join("Warp")
                    .join("config")
                    .join("cli")
            })
            .unwrap_or_else(|| {
                home()
                    .join("AppData")
                    .join("Local")
                    .join("warp")
                    .join("Warp")
                    .join("config")
                    .join("cli")
            });
        WarpCliPaths {
            settings: dir.join("settings.toml"),
            mcp_config: dir.join(".mcp.json"),
            dir,
        }
    } else if cfg!(target_os = "macos") {
        let dir = home().join(".warp_cli");
        WarpCliPaths {
            settings: dir.join("settings.toml"),
            mcp_config: dir.join(".mcp.json"),
            dir,
        }
    } else {
        let dir = xdg_config_home().join("warp-terminal").join("cli");
        WarpCliPaths {
            settings: dir.join("settings.toml"),
            mcp_config: dir.join(".mcp.json"),
            dir,
        }
    }
}

/// Every MCP file TokSave manages for Warp: legacy desktop, official desktop,
/// platform CLI, and the documented `~/.warp_cli/.mcp.json` path.
pub fn warp_mcp_files() -> Vec<PathBuf> {
    let p = warp_paths();
    let cli = warp_cli_paths();
    let mut files = vec![p.mcp_config, p.mcp_config_official, cli.mcp_config];
    let documented = home().join(".warp_cli").join(".mcp.json");
    if !files.contains(&documented) {
        files.push(documented);
    }
    files
}

pub fn warp_known_bin_dirs() -> Vec<PathBuf> {
    vec![home().join(".local").join("bin")]
}

pub fn warp_desktop_paths() -> Vec<PathBuf> {
    if cfg!(windows) {
        if let Some(local) = env::var_os("LOCALAPPDATA") {
            return vec![
                PathBuf::from(local)
                    .join("Programs")
                    .join("Warp")
                    .join("Warp.exe"),
            ];
        }
        vec![]
    } else if cfg!(target_os = "macos") {
        vec![PathBuf::from("/Applications/Warp.app")]
    } else {
        vec![PathBuf::from("/usr/bin/warp-terminal")]
    }
}

pub fn local_bin() -> PathBuf {
    if cfg!(windows) {
        if let Some(la) = env::var_os("LOCALAPPDATA") {
            return PathBuf::from(la).join("Programs").join("toksave");
        }
        home()
            .join("AppData")
            .join("Local")
            .join("Programs")
            .join("toksave")
    } else {
        home().join(".local").join("bin")
    }
}

pub fn cache_dir() -> PathBuf {
    if let Some(c) = env::var_os("TOKSAVE_CACHE_DIR") {
        return PathBuf::from(c);
    }
    home().join(".cache").join("toksave")
}

pub fn ensure_dir(p: &Path) -> Result<()> {
    if !p.exists() {
        fs::create_dir_all(p)?;
    }
    Ok(())
}

pub fn read_file(p: &Path) -> Option<String> {
    fs::read_to_string(p).ok()
}

pub fn write_file(p: &Path, content: &str) -> Result<()> {
    if let Some(parent) = p.parent() {
        ensure_dir(parent)?;
    }
    let pid = std::process::id();
    let tmp = p.with_extension(format!(
        "{}.{}.tmp",
        p.extension().unwrap_or_default().to_string_lossy(),
        pid
    ));
    let mut f = fs::File::create(&tmp)?;
    f.write_all(content.as_bytes())?;
    f.flush()?;
    fs::rename(&tmp, p)?;
    Ok(())
}

pub fn toksave_abs() -> String {
    std::env::current_exe()
        .map(|p| {
            let s = p.to_string_lossy().to_string();
            // Forward slashes work in cmd.exe, PowerShell, and Git Bash;
            // backslashes break Git Bash hooks (see #24).
            if cfg!(windows) {
                s.replace('\\', "/")
            } else {
                s
            }
        })
        .unwrap_or_else(|_| "toksave".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn home_uses_env() {
        let _g = crate::util::env_test_lock();
        let tmp = env::temp_dir().join("toksave-paths-test-home");
        fs::create_dir_all(&tmp).unwrap();
        let old = env::var_os("HOME");
        // Safe: serialized by env_test_lock against other env-mutating tests.
        unsafe {
            env::set_var("HOME", &tmp);
        }
        assert_eq!(home(), tmp);
        unsafe {
            if let Some(o) = old {
                env::set_var("HOME", o);
            } else {
                env::remove_var("HOME");
            }
        }
    }

    #[test]
    fn cache_dir_uses_env_override() {
        let _g = crate::util::env_test_lock();
        let tmp = env::temp_dir().join("toksave-cache-test");
        let old = env::var_os("TOKSAVE_CACHE_DIR");
        // Safe: serialized by env_test_lock against other env-mutating tests.
        unsafe {
            env::set_var("TOKSAVE_CACHE_DIR", &tmp);
        }
        assert_eq!(cache_dir(), tmp);
        unsafe {
            if let Some(o) = old {
                env::set_var("TOKSAVE_CACHE_DIR", o);
            } else {
                env::remove_var("TOKSAVE_CACHE_DIR");
            }
        }
    }

    #[test]
    fn write_file_is_atomic_and_readable() {
        let p = env::temp_dir().join("toksave-write-test.txt");
        write_file(&p, "hello\n").unwrap();
        assert_eq!(read_file(&p).as_deref(), Some("hello\n"));
        fs::remove_file(&p).ok();
    }

    #[test]
    fn claude_paths_under_home() {
        let _g = crate::util::env_test_lock();
        let tmp = env::temp_dir().join("toksave-claude-test");
        let old = env::var_os("HOME");
        // Safe: serialized by env_test_lock against other env-mutating tests.
        unsafe {
            env::set_var("HOME", &tmp);
        }
        let cp = claude_paths();
        assert_eq!(cp.dir, tmp.join(".claude"));
        assert_eq!(cp.global_json, tmp.join(".claude.json"));
        unsafe {
            if let Some(o) = old {
                env::set_var("HOME", o);
            } else {
                env::remove_var("HOME");
            }
        }
    }

    #[test]
    fn warp_paths_include_official_and_cli_mcp() {
        let _g = crate::util::env_test_lock();
        let tmp = env::temp_dir().join("toksave-warp-paths-test");
        fs::create_dir_all(&tmp).unwrap();
        let old_home = env::var_os("HOME");
        let old_xdg = env::var_os("XDG_CONFIG_HOME");
        let old_local = env::var_os("LOCALAPPDATA");
        unsafe {
            env::set_var("HOME", &tmp);
            env::set_var("XDG_CONFIG_HOME", tmp.join("xdg"));
            env::set_var("LOCALAPPDATA", tmp.join("Local"));
        }
        let p = warp_paths();
        assert_eq!(p.mcp_config, tmp.join(".warp").join("mcp.json"));
        assert_eq!(p.mcp_config_official, tmp.join(".warp").join(".mcp.json"));
        let cli = warp_cli_paths();
        assert_eq!(cli.mcp_config.file_name().unwrap(), ".mcp.json");
        let files = warp_mcp_files();
        assert!(files.contains(&p.mcp_config));
        assert!(files.contains(&p.mcp_config_official));
        assert!(files.contains(&cli.mcp_config));
        assert!(files.len() >= 3);
        unsafe {
            match old_home {
                Some(o) => env::set_var("HOME", o),
                None => env::remove_var("HOME"),
            }
            match old_xdg {
                Some(o) => env::set_var("XDG_CONFIG_HOME", o),
                None => env::remove_var("XDG_CONFIG_HOME"),
            }
            match old_local {
                Some(o) => env::set_var("LOCALAPPDATA", o),
                None => env::remove_var("LOCALAPPDATA"),
            }
        }
    }
}
